src/i18n.ts

```diff
@@ -0,0 +1,316 @@
+import {
+  Context,
+  Fluent,
+  type FluentBundleOptions,
+  FluentType,
+  type FluentValue,
+  type LocaleId,
+  type MiddlewareFn,
+  type NextFunction,
+  resolve,
+  type TranslationContext,
+} from "./deps.ts";
+
+import { readLocalesDir, readLocalesDirSync } from "./utils.ts";
+
+import type { I18nConfig, I18nFlavor, TranslateFunction } from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  // deno-lint-ignore no-explicit-any
+  toString(scope: any): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom Fluent function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
+  if (args[1] && typeof args[1] !== "string") {
+    throw new TypeError("CTX: Only string type is allowed for defaultValue.");
+  }
+
+  const key = args[0];
+  const defaultValue = typeof args[1] === "string" ? args[1] : undefined;
+
+  if (typeof key === "string") {
+    return new FluentContext(key, defaultValue);
+  }
+
+  throw new TypeError(
+    "CTX: Only type string is allowed for key.",
+  );
+}
+
+export class I18n<C extends Context = Context> {
+  private config: I18nConfig<C>;
+  readonly fluent: Fluent;
+  readonly locales = new Array<string>();
+
+  constructor(config: Partial<I18nConfig>) {
+    this.config = {
+      defaultLocale: "en",
+      fluentBundleOptions: { functions: {} },
+      ...config,
+    };
+
+    if (
+      this.config.fluentBundleOptions &&
+      this.config.fluentBundleOptions.functions
+    ) {
+      this.config.fluentBundleOptions.functions["CTX"] = CTX;
+    } else {
+      this.config.fluentBundleOptions = {
+        ...this.config.fluentBundleOptions,
+        functions: { CTX },
+      };
+    }
+
+    this.fluent = new Fluent(this.config.fluentOptions);
```

Comment:
I got this error while test it on Windows10

```
$ deno task dev
Warning deno task is unstable and may drastically change in the future
Task dev export STAGE=dev; deno run --allow-env --allow-read --allow-write --allow-net poll.ts
Check file:///S:/Projects/bot-vacation/poll.ts
Error in middleware while handling update 834302471 TypeError: Cannot read properties of undefined (reading 'fluent')
    at middleware (file:///S:/Projects/bot-vacation/pkg/grammy-i18n/src/i18n.ts:190:28)
    at https://deno.land/x/grammy@v1.9.2/composer.ts:126:41
    at https://deno.land/x/grammy@v1.9.2/composer.ts:137:19
    at pass (https://deno.land/x/grammy@v1.9.2/composer.ts:142:12)
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
    at https://deno.land/x/grammy@v1.9.2/composer.ts:134:15
No error handler was set!
Set your own error handler with `bot.catch = ...`
Stopping bot
error: Uncaught (in promise) BotError: TypeError in middleware: Cannot read properties of undefined (reading 'fluent')
```

At first I thought it was my plugin fault but even if I disable it. It's still the same error. (I'm investing on it)

Reply:
I have a bot running on production in Deno using this plugin, no issues like this. Are you sure your local plugin repo is up to date?

Reply:
I accidentally reload the page and lost all the detail comment... Sorry, but I will be super short on this:

Here is what I get from console.log

```log
(1) { directory: "./src/l10n" }
(2) {
  defaultLocale: "en",
  fluentBundleOptions: { functions: {} },
  directory: "./src/l10n"
}
(3) undefined
(4) Fluent {
  options: {},
  bundles: Set {},
  defaultBundle: undefined,
  warningHandler: LoggingWarningHandler { options: {}, writeLog: [Function: bound callConsole] }
}
```

In I18n class constructor:
1. config
1. this.config
1. this.config.fluentOptions
1. this.fluent



I use non-session setup:

```ts
import { Bot, session } from 'grammy'
import { I18n } from "grammy/i18n"

const bot = new Bot<SyntheticContext>(BOT_TOKEN)
bot.catch(console.error)

const i18n = new I18n({
	directory: './src/l10n'
})
bot.use(i18n.middleware)

bot.command('x', async (ctx) =>
{
	await ctx.reply(ctx.t('cmd-x-hi'))
})
```

Reply:
I forgot, here is my imported map (deps.ts to other people)

```json
{
	"imports": {
		"deno/": "https://deno.land/std@0.149.0/",
		"grammy": "https://deno.land/x/grammy@v1.9.2/mod.ts",
		"grammy/i18n": "./pkg/grammy-i18n/src/mod.ts",
		"denodb": "https://raw.githubusercontent.com/joeldesante/denodb/master/mod.ts",
		"denodb/": "https://raw.githubusercontent.com/joeldesante/denodb/master/lib/",
		"sift": "https://deno.land/x/sift@0.5.0/mod.ts",
		"@lib/": "./src/lib/",
		"@res/": "./src/res/",
		"#cfg": "./config.ts",
		"#def": "./src/define.ts"
	}
}
```

Reply:
> bot.use(i18n.middleware)

I don't think that will work. You should either do,

```ts
// grammY calls middleware() internally.
bot.use(i18n);
```

or call the middleware function.

```ts
bot.use(i18n.middleware());
```

Reply:
>     bot.use(i18n.middleware());

I thought so but vscode said that is not possible. And when I tried to run it, I got this:

```log
error: TS2345 [ERROR]: Argument of type 'MiddlewareFn<Context & I18nFlavor>' is not assignable to parameter of type 'Middleware<SyntheticContext>'.
  Type 'MiddlewareFn<Context & I18nFlavor>' is not assignable to type 'MiddlewareFn<SyntheticContext>'.
    Types of parameters 'ctx' and 'ctx' are incompatible.
      Type 'SyntheticContext' is not assignable to type 'Context & I18nFlavor'.
        Type 'SyntheticContext' is not assignable to type 'Context'.
          The types of 'api.raw.setWebhook' are incompatible between these types.
            Type '(args: { url: string; certificate?: InputFile | undefined; ip_address?: string | undefined; max_connections?: number | undefined; allowed_updates?: readonly ("message" | "edited_message" | ... 11 more ... | "chat_join_request")[] | undefined; drop_pending_updates?: boolean | undefined; secret_token?: string | undefi...' is not assignable to type '(args: { url: string; certificate?: InputFile | undefined; ip_address?: string | undefined; max_connections?: number | undefined; allowed_updates?: readonly ("message" | "edited_message" | ... 11 more ... | "chat_join_request")[] | undefined; drop_pending_updates?: boolean | undefined; }, signal?: AbortSignal | unde...'.    
              Types of parameters 'args' and 'args' are incompatible.
                Type '{ url: string; certificate?: InputFile | undefined; ip_address?: string | undefined; max_connections?: number | undefined; allowed_updates?: readonly ("message" | "edited_message" | ... 11 more ... | "chat_join_request")[] | undefined; drop_pending_updates?: boolean | undefined; }' is not assignable to type '{ url: string; certificate?: InputFile | undefined; ip_address?: string | undefined; max_connections?: number | undefined; allowed_updates?: readonly ("message" | "edited_message" | ... 11 more ... | "chat_join_request")[] | undefined; drop_pending_updates?: boolean | undefined; secret_token?: string | undefined; }'.
                  Types of property 'certificate' are incompatible.
                    Type 'import("https://deno.land/x/grammy@v1.8.3/platform.deno.ts").InputFile | undefined' is not assignable to type 'import("https://deno.land/x/grammy@v1.9.2/platform.deno.ts").InputFile | undefined'.
                      Property '[toRaw]' is missing in type 'import("https://deno.land/x/grammy@v1.8.3/platform.deno.ts").InputFile' but required in type 'import("https://deno.land/x/grammy@v1.9.2/platform.deno.ts").InputFile'.    
bot.use(i18n.middleware())
        ~~~~~~~~~~~~~~~~~
    at file:///S:/Projects/bot-vacation/src/bot.ts:37:9

    '[toRaw]' is declared here.
        async [toRaw](): Promise<
              ~~~~~~~
        at https://deno.land/x/grammy@v1.9.2/platform.deno.ts:105:11
```


About the context flavour, here is mine: 

```ts
export interface DatabaseFlavour
{
	db: DatabaseMethodFlavour
}

export type SyntheticContext = Context & DatabaseFlavour & I18nFlavor
```

Reply:
You have version mismatch. v1.8.3 and v1.9.2. So, this is almost unrelated to this plugin. 

`Type 'import("https://deno.land/x/grammy@v1.8.3/platform.deno.ts").InputFile | undefined' is not assignable to type 'import("https://deno.land/x/grammy@v1.9.2/platform.deno.ts").InputFile | undefined'.`

And `bot.use(i18n)` or `bot.use(i18n.middleware())` is the correct usage of it. Never do `i18n.middleware`.

Reply:
You're VSCode is wrong. Try restarting it after fixing the version mismatch. Also is there a specific reason to stick with grammY 1.9.2? Update to 1.10.x, and do a forced `--reload` of cached dependencies. That should fix you're type issues. From the logs, I am pretty sure that your `lib.deno.dev/x/grammy` is still in v1.8.3 (or 1.9.2). Thats what causing this.